### ISSUE TITLE
通过强化学习优化交易逻辑

### ISSUE THREAD

**Issue by chenditc** [chenditc/basic_trading_strategy#28]
目前交易逻辑是固定的预设值，例如 topkdropout，只持有 topk 只股票，每日限定换仓股票不超过 N 只。

由于预测模型的不同，最优的 topk 和 N 值会有所区别。尝试用强化学习优化交易的逻辑：

---

建模1:

简化可选的 action，使得采样效率更高。

环境初始化：
 - 使用一半日期的数据集进行训练，训练后导出剩余时间的预测数据。
 - 将所有交易日的价格数据、预测数据打包为环境包。
 - 设置初始交易金额

环境状态：
 - top 300 股票的预测值
 - 当前股票是否持有

可选动作：
 - 最大换仓数 N，N <= K 

状态转移：
 - 每日交易top k的股票作为持仓。
 - 根据每日输出的 N 不同，调整换仓。
 - 根据每日的价格数据返回持仓收益变化情况。
 - 完整走完训练集，则为一次完整的 epoch

期望：
 - 比固定的 topk drop out 表现更好或者一致。

---

建模2:

展开动作空间，尝试获取更强的结果。从 M 个股票中选择 N 只股票。

动作方案1：
 - action space 有 N 个纬度，每个维度是一个有 M 个选择的 discrete space。

动作方案2:
 - action space 为 M 维的连续空间，然后取最大的 N 个股票作为实际的交易决策。

动作方案3:
- 神经网络最后一层为 M 个节点，用 MultiNominal 是不是可以取出 N 个 在 1-M 之间的数字？

动作方案4:
- 每个股票作为一个 Agent，输出1-10 的 ranking，选ranking 最大的agent 买入。

动作方案5:
- 每一次输出一个M维的discrete action，然后重复N次，当选满 N 个股票后再模拟交易并计算收益。由于使用的是 GAE 计算 Advantage，当 lambda = 1 时，与同时选 N 个股票的效果是一样的。

**Comment by chenditc** [comment]
一些思考：
1. 怎么表示 action space 和 state space。
  - 看了一些发布出来的建模方法，都是以固定股票池作为 action space 和 state space，每个股票固定对应一个 action space 里的output。但是这个方法可能会导致模型学会买某 “个” 股票，而不是根据股票的特点分析后买股票。例如模型可能会学会买爆 MSFT 或者买爆茅台，即在最后一层的输出层，无论输入是什么，都激活茅台对应的输出点。
  - 把所有股票放在一个输出集合的目的是为了实现 “选择” 这个操作。但是神经网络模型的输出节点数是固定的，不能适应在不断变化的股票池。

**Comment by chenditc** [comment]
先实现了一个版本，测试强化学习的拟合学习能力。用测试集的信号进行训练，并尝试优化回测结果。

1. 以 day 2 open label 训练，以 top n=50，drop k=5进行回测：
```
'The following are analysis results of benchmark return(1day).'
                       risk
mean               0.000477
std                0.012295
annualized_return  0.113561
information_ratio  0.598699
max_drawdown      -0.370479
'The following are analysis results of the excess return without cost(1day).'
                       risk
mean               0.000224
std                0.013000
annualized_return  0.053239
information_ratio  0.265467
max_drawdown      -0.351031
'The following are analysis results of the excess return with cost(1day).'
                       risk
mean               0.000132
std                0.012998
annualized_return  0.031356
information_ratio  0.156370
max_drawdown      -0.363148
'The following are analysis results of indicators(1day).'
     value
ffr    1.0
pa     0.0
pos    0.0
```

**Comment by chenditc** [comment]
目前的实现：https://github.com/chenditc/basic_trading_strategy/tree/main/strategies/rl/topkdropout_gym

**Comment by chenditc** [comment]
对比强化学习的 N dropout 效果：
**奇怪的是 baseline 结果和训练时回测的不一样，不知道会不会是训练时用的 close price 计算**

### 4d prediction
**Baseline: Top k =50, N dropout = 50**
```
'The following are analysis results of benchmark return(1day).'
                       risk
mean               0.000477
std                0.012295
annualized_return  0.113561
information_ratio  0.598699
max_drawdown      -0.370479
'The following are analysis results of the excess return without cost(1day).'
                       risk
mean               0.002530
std                0.013046
annualized_return  0.602189
information_ratio  2.991980
max_drawdown      -0.166997
'The following are analysis results of the excess return with cost(1day).'
                       risk
mean               0.001736
std                0.013033
annualized_return  0.413084
information_ratio  2.054452
max_drawdown      -0.178446
```

**RL result : Top k =50, N dropout = 50**

```
'The following are analysis results of the excess return without cost(1day).'
                       risk
mean               0.002392
std                0.013075
annualized_return  0.569266
information_ratio  2.822107
max_drawdown      -0.174559
'The following are analysis results of the excess return with cost(1day).'
                       risk
mean               0.001676
std                0.013073
annualized_return  0.398856
information_ratio  1.977622
max_drawdown      -0.185889
```
### 1d prediction
**Baseline: Top k =50, N dropout = 50**
```
'The following are analysis results of benchmark return(1day).'
                       risk
mean               0.000477
std                0.012295
annualized_return  0.113561
information_ratio  0.598699
max_drawdown      -0.370479
'The following are analysis results of the excess return without cost(1day).'
                       risk
mean               0.002223
std                0.012698
annualized_return  0.529067
information_ratio  2.700814
max_drawdown      -0.158160
'The following are analysis results of the excess return with cost(1day).'
                       risk
mean               0.001240
std                0.012690
annualized_return  0.295126
information_ratio  1.507439
max_drawdown      -0.170191
```

**Comment by chenditc** [comment]
4d 达不到 baseline 的可能性：
 - 采样的 reward 不是即时 action 的 reward，当前的选择会在 T+1 展现出来

训练时 baseline 的 reward 是 200，训练出来的是 192 左右，训练期间达到过 199.

**Comment by chenditc** [comment]
下一步：
1. 整理一下已经实现的代码，把可以复用的 simulator 抽出来。
2. 实现新的 action interpreter，直接进行选股。

**Comment by chenditc** [comment]
![rldeep drawio](https://user-images.githubusercontent.com/3244845/219867113-4606c39d-0e52-4fa7-b3dd-46c834d2f529.png)

**Comment by chenditc** [comment]
训练的曲线很奇怪，在接近200之后，有可能会回落。是因为训练步长超过 Trust zone 吗？

![image](https://user-images.githubusercontent.com/3244845/219991453-8d3cedb9-23de-4b96-a7a7-740049c7ae5a.png)

**Comment by chenditc** [comment]
![image](https://user-images.githubusercontent.com/3244845/220514454-d6f055ea-255b-4139-8307-e82b8e4b6224.png)
[pm_rl.zip](https://github.com/chenditc/basic_trading_strategy/files/10799773/pm_rl.zip)

**Comment by chenditc** [comment]
实现了一个版本。在之前的思路上做了一些额外的修改：

1. 决策模型的变更。如果需要选 N 只股票出来，则将其分为 N 个决策步骤，每个步骤选出第 X 个已持仓的股票，与剩余股票池的股票放在一起进行选择。
  - 这样做的思路是来源于 RankNet 的 pairwise 排序。选股的目的和搜索引擎排序的目的是相似的。通过把决策拆散到每一个股票的选择上，我们就可以得到更多的训练标签。
2. 对股票池进行采样。第一次采样选择每一次轨迹可以进行交易的股票，这样可以变换股票池，获得更多排列组合，而不是局限于预测信号最好的 top 股票。因为神经网络的输入 shape 是固定的，但并不是所有股票在当日都能交易或者都能生成预测信号，所以第二次采样则是在股票池的基础上进行 truncate 裁剪，只取前 N 个股票，保证输入的完整性。这样其实是 listwise 的排序训练，当 N = 2 时，就简化成了 pairwise 的训练了。
3. 过滤掉第二天不可交易的股票。避免回测时由于无法交易，实际选择的动作和收益并不直接关联，导致收益有噪声。
4. 使用步骤2中裁剪剩下的股票池的中值回报作为 benchmark，用来计算超额收益。因为我们对股票池进行过了采样，所以使用指数作为 benchmark 的超额收益不能很好地表示我们选择的好坏，用裁剪剩下的股票池更好。使用中值而不用均值是为了避免极值导致的 benchmark 预估失真。
5. 训练时只持仓1只股票。如果持仓数量较大，例如持有20只股票，那么换仓的动作好坏可能不明显，奖励值波动不由动作决定，收敛可能更慢。

目前还不能很好地收敛。考虑修改：
1. GAE 的参数。由于当前动作对未来的影响可能主要在于第二天，而不是很久以后的第10天，所以可以考虑修改 GAE 参数，使得反馈更快。同时使更久以后的收益影响更小一些。

**Comment by chenditc** [comment]
我主要参考了一下这篇博客：https://blog.csdn.net/weixin_39891381/article/details/105153867

PPO 的 GAE reward 相关参数有两个：
1. GAE lambda。这个控制了当前状态开始，有几个 episode 的 advantage 一起被考虑进去。当 lambda = 0 时，相当于 TD(0)，当 lambda = 1 时，相当于 MC。
2. Gamma。这个控制了Advantage 有多少由 reward 计算，有多少由 V 函数进行估计。Gamma = 1 时，相当于全部由 V 函数进行估计，Gamma = 0 时，相当于全部由 reward 估计。所以当 Gamma 越小时，偏差越小，方差越大。

**Comment by chenditc** [comment]
当前：交易手续费开仓千分之一，平仓千分之二的情况下，以沪深300作为股票池，topkdropout 100% 换手的情况下作为 baseline，baseline reward 是 11.9，强化学习能达到 12-13 左右。

下一步：
1. 调优 simulator，使得采样速度更快，目前不知道什么原因，一个 iteration 需要10分钟。
2. 修改 action space，不对股票进行排序，只提供股票的预测值，看是否也能收敛到同样的水平。
3. 提供更多 feature，看是否能有更好的效果。

**Comment by chenditc** [comment]
尝试实现的 MARL 方案：

方案1：在 observation 上增加一个维度，表示 num_agents，action 也同样增加该维度。这样相当于所有 agent 在同一时刻做出 action，即给出当前的 action vote。这样实现的好处在于可以在同一时间给出 reward，不需要每个 agent 做动作时都给 reward。但是 update 的过程中发现两个问题：
1. reward、indices 仍然是单维度的。reward 在框架中只接受 float 类型，不接受 numpy 类型。
2. indices 表示当前更新的 minibatch 在整个数据集中的索引，每一个 step 会增加一个 indices，这个会在后续用来 filter termination flag 和 v_s。如果要用这个方案，则需要修改 tianshou 的 policy 的 compute reward 或者 update 的实现。

方案2：实现一个 petting zoo 的 ACE 环境

**Comment by chenditc** [comment]
目前用方案1 实现了一个 MARL 的方案。但是训练效果不好，原因大概有这么几个：
1. 奖励信号的方差和随机性大，相同选择下，智能体获取到的奖励不稳定。所以 PPO 策略中利用势能来优化收益计算的效果不好，GAE 的 lambda 较高时收敛难，较低时收敛快。
2. 多智能体的输入信号构造难以做到同分布。目前的做法是把时间段分成两个，分别训练两个模型，然后预测没有加入训练集的时间，再把预测信号拼接起来。这样拼接的信号和最终用来输出的测试集的信号特征似乎也不太一样。训练过程中训练集的表现在不断提升，但是验证集的表现在不断下降。
3. 多智能体的 reward 信号相同，不能很好区分每个智能体对 reward 的贡献值。

**Comment by chenditc** [comment]
总结一下过去一个月的结果：

**问题的设置：**
  - 时间段划分：05-08年验证集，08-16年训练集，17-19年测试集
  - 强化学习的数据准备：
  1. 训练模型，模型在08-13年、13-16年、08-16年分别训练一次，以验证集作为 early stopping。
  2. 导出测试数据，08-13年的模型导出13-16年的预测 signal，13-16年的模型导出08-13年的 signal，08-16年的模型导出17-19年的 signal
  3. 使用不同的模型作1、2步骤，生成 M 个 signal，以这 M 个 signal + 当前是否持仓（1/0）作为股票数据集的输入 （M+1 维)
  4. 相当于学习一个 ensemble 的模型，同时把持仓信息考虑进去。
  - Baseline: 以表现最好的模型来选择持仓。

**尝试过的事情：**
1. 使用排序后的股票减少动作空间的方法确实是可行的，也可以训练到收敛，但是当输入的 feature 变多后，没有很好的排序标准，所以这个方法后续没有继续使用。
2. 训练时采样不同的股票池确实可以提高泛化能力，可以在验证集上达到较稳定的效果。
3. 使用 MARL 的方式训练时，手动修改了一下网络来实现 MARL，每个股票的 feature 输出一个 logits，然后对最后的 logits 合并后用 Categorical distribution 来 Sample 最后选择的股票。 这样的方式相当于重新对股票做了一个 ranking，然后选择 topk 的股票进行交易。每一次 Sample 选择一个股票，重复 N 次就选出了 N 只股票。
5. 使用默认的 PPO 算法很难收敛。后来发现可能问题在于股票数据的随机性非常强，随机性指的是智能体的输入对于环境的影响小，而环境本身对收益的随机影响较大。PPO 算法依赖 critic network 来学习可靠的 value function 从而引导梯度，但是由于随机性强，value function 很难预测某一个状态的预期收益是多少，也就很难收敛。所以修改了 lambda 为比较小的值 0 - 0.5，同时设置 vf_coef = 0，即在计算 loss 时不计算 value function 的梯度，就可以比较好地收敛了。

**结果：**
在训练集、验证集上可以超越 baseline，但是在测试集上并没有达到，在最好的 signal 和最差的 signal 之间。
验证集 baseline 的超额收益大约在 1.6， rl 算法能达到 2.8 - 3.1。
测试集的 baseline 的超额收益在 0.99 - 1.9 之间，rl 算法大约在 1.2 - 1.3.

**可能原因：**
1. 过去的 Pattern 和当前的不一致，导致过去最优的 trajectory 并不是当前最优的 trajectory。
2. 输入的信息不够泛化，由于只输入了 signal 的信息，所以容易根据 signal 的 pattern 过拟合。

**后续：**
1. 既然 PPO 的 critic network 对任务没有很大的帮助，可能使用 DQN family 的算法会更合适，更容易收敛。
2. 目前使用 lambda = 0 相当于不考虑未来收益，也就没有把长期的轨迹考虑进去，所以没有发挥 RL 的强项。
3. 不同时间段之间，可能 signal 预测的准确度并不相同，这部分也是影响泛化能力的一个问题。相当于我们在 ensemble 阶段过拟合了。这个应该是 ensemble 方向已经解决过的问题，比如添加 L1 L2 的 regularization？还有其他的方法吗？
4. 问题本身的设置导致 RL 并不会关注持仓信息，并不会产生换手率更低的策略。目前设置的是万二佣金，千1印花税。收益的来源可能 95% 是由 signal 的准确性贡献的，而换仓的手续费可能只会影响 5% 的收益。所以 RL 会更多考虑优化 ensemble 的信息来产生更多的收益。
